src/Interpreters/JIT/CHJIT.cpp

```diff
@@ -0,0 +1,388 @@
+    module_info.identifier = current_module_key;
+
+    module_identifier_to_memory_manager[current_module_key] = std::move(module_memory_manager);
+
+    compiled_code_size.fetch_add(module_info.size, std::memory_order_relaxed);
+
+    return module_info;
+}
+
+void CHJIT::deleteCompiledModule(const CHJIT::CompiledModuleInfo & module_info)
+{
+    std::lock_guard<std::mutex> lock(jit_lock);
+
+    auto module_it = module_identifier_to_memory_manager.find(module_info.identifier);
+    if (module_it == module_identifier_to_memory_manager.end())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "There is no compiled module with identifier {}", module_info.identifier);
+
+    for (const auto & function : module_info.compiled_functions)
+        name_to_symbol.erase(function);
+
+    module_identifier_to_memory_manager.erase(module_it);
+    compiled_code_size.fetch_sub(module_info.size, std::memory_order_relaxed);
+}
+
+void * CHJIT::findCompiledFunction(const CompiledModuleInfo & module_info, const std::string & function_name) const
+{
+    std::lock_guard<std::mutex> lock(jit_lock);
+
+    std::string symbol_name = std::to_string(module_info.identifier) + '_' + function_name;
+    auto it = name_to_symbol.find(symbol_name);
+    if (it != name_to_symbol.end())
+        return it->second;
+
+    return nullptr;
+}
+
+void CHJIT::registerExternalSymbol(const std::string & symbol_name, void * address)
+{
+    std::lock_guard<std::mutex> lock(jit_lock);
+    symbol_resolver->registerSymbol(symbol_name, address);
+}
+
+std::string CHJIT::getMangledName(const std::string & name_to_mangle) const
+{
+    std::string mangled_name;
+    llvm::raw_string_ostream mangled_name_stream(mangled_name);
+    llvm::Mangler::getNameWithPrefix(mangled_name_stream, name_to_mangle, layout);
+    mangled_name_stream.flush();
+
+    return mangled_name;
+}
+
+void CHJIT::runOptimizationPassesOnModule(llvm::Module & module) const
+{
+    llvm::PassManagerBuilder pass_manager_builder;
+    llvm::legacy::PassManager mpm;
+    llvm::legacy::FunctionPassManager fpm(&module);
+    pass_manager_builder.OptLevel = 3;
+    pass_manager_builder.SLPVectorize = true;
+    pass_manager_builder.LoopVectorize = true;
+    pass_manager_builder.RerollLoops = true;
+    pass_manager_builder.VerifyInput = true;
+    pass_manager_builder.VerifyOutput = true;
+    machine->adjustPassManager(pass_manager_builder);
+
+    fpm.add(llvm::createTargetTransformInfoWrapperPass(machine->getTargetIRAnalysis()));
+    mpm.add(llvm::createTargetTransformInfoWrapperPass(machine->getTargetIRAnalysis()));
+
+    pass_manager_builder.populateFunctionPassManager(fpm);
+    pass_manager_builder.populateModulePassManager(mpm);
+
+    fpm.doInitialization();
+    for (auto & function : module)
+        fpm.run(function);
+    fpm.doFinalization();
+
+    mpm.run(module);
+}
+
+std::atomic<bool> initialized = false;
+
+static void initializeLLVMTarget()
+{
+    if (initialized)
```

Comment:
It's also can be done with a struct with constructor and function level static var.
Also with std::once_flag.

Reply:
@alexey-milovidov did not want to use struct with constructor, because static object construction is unspecified, and client may use ```static CHJIT jit```. std::once_flag can be used.